gbus/metrics/handler_metrics.go

```diff
@@ -0,0 +1,131 @@
+package metrics
+
+import (
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_model/go"
+	"github.com/sirupsen/logrus"
+	"sync"
+)
+
+var (
+	handlerMetricsByHandlerName  = make(map[string]*HandlerMetrics)
+	lock      					 = &sync.Mutex{}
+)
+
+const (
+	Failure           = "failure"
+	Success           = "success"
+	ExceededRetries   = "exceeded_retries"
+	HandlerResult     = "result"
+	GrabbitNamespace  = "grabbit"
+	HandlersSubsystem = "handlers"
+)
+
+type HandlerMetrics struct {
+	result  *prometheus.CounterVec
+	latency prometheus.Summary
+}
+
+func AddHandlerMetrics(handlerName string) {
+	lock.Lock()
+	defer lock.Unlock()
+	_, ok := handlerMetricsByHandlerName[handlerName]
+	if !ok {
+		handlerMetricsByHandlerName[handlerName] = newHandlerMetrics(handlerName)
+	}
+}
+
+func RunHandlerWithMetric(handleMessage func() error, handlerName string, logger logrus.FieldLogger) error {
+	handlerMetrics, ok := handlerMetricsByHandlerName[handlerName]
+
+	if !ok {
+		logger.WithField("handler", handlerName).Warn("Running with metrics - couldn't find metrics for the given handler")
+		return handleMessage()
+	}
+
+	err := trackTime(handleMessage, handlerMetrics.latency)
+
+	if err != nil {
+		handlerMetrics.result.WithLabelValues(Failure).Inc()
+	} else {
+		handlerMetrics.result.WithLabelValues(Success).Inc()
+	}
+
+	return err
+}
+
+func ReportHandlerExceededMaxRetries(handlerName string, logger logrus.FieldLogger) {
```

Comment:
I do not think this metric should be reported on handlers but rather on rejected messages.
In the current setup and with the above code if a message s rejected it may result in multiple metrics reported for the same message which seem wrong.
think that this should be in a message_metrics.go and named ReportRejectedMessage.
@danielwitz,  @vladshub WDYT ?

Reply:
regardless the metric for failed message handling might happen multiple times for the same message, didn't find a way around that.
In the current setup if the handler exceeds the retries count the metric failure will be increased by that count and we'll have increased the exceeded retries metric only once.
Rejected message might be a better name though

Reply:
the handler can indeed be invoked more than once that is why the rejected message metric should only be reported (once) after all retries fail.